### PR TITLE
Subgroup fixes

### PR DESCRIPTION
**Pull request by cdavis5e**
Assorted fixes to subgroup support. This gets most of the subgroup tests passing for compute shaders. The remaining subgroup tests fail either because the graphics tests use isoline tessellation, or because of bugs in Metal.

### CODE REVIEW

File: spirv_msl.cpp [hunk 1]
```diff
@@ -4610,6 +4610,89 @@ void CompilerMSL::emit_custom_functions()
...
+			statement("inline T spvSubgroupBroadcast(T value, ushort lane)");
+			begin_scope();
+			if (msl_options.is_ios())
+				statement("return quad_broadcast(value, lane);");
+			else
+				statement("return simd_broadcast(value, lane);");
+			end_scope();
+			statement("");
+			statement("template<>");
+			statement("inline bool spvSubgroupBroadcast(bool value, ushort lane)");
```
Comment (HansKristian-Work) [inline spirv_msl.cpp:4626]: Can we make this more compact? Either specialized template on boolean types, or emitting code in a loop.

**Review by HansKristian-Work — CHANGES_REQUESTED** [review]
Nice! Yes, I figured we'd need something like CTS to properly test subgroups ...

My only real concern is the intense code bloat to stamp out variants. Please try to find a way to make the MSL codegen less repetitive.

File: spirv_msl.cpp (hunk 1 above)
Comment (cdavis5e) [reply spirv_msl.cpp:4626]: I didn't even think TP inference with a non-type parameter would work. But it worked pretty well, in fact.

**Comment by cdavis5e** [comment]
I simplified the `SubgroupG[et]Mask` changes a bit with `insert_bits()`. This should make it a bit faster and easier to comprehend.

**Review by HansKristian-Work — APPROVED** [review]
LGTM.

**Pull request merged**